### DISCUSSION TITLE
How to handle range for non-trivial keys?

### DISCUSSION THREAD

**Discussion by jamesmunns** [fjall-rs/fjall#106] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**

It is unclear to me how [`range()`](https://docs.rs/fjall/latest/fjall/struct.ReadTransaction.html#method.range) could work with non trivial keys.

For example, if I wanted to iterate over the prefixed range:

`devices/123/logs/000000AF` to `devices/123/logs/000000BD`, I can't use these as ranges. I did try implementing `RangeBounds` for this, but it doesn't seem to work as I think, because the containing type needs to impl AsRef, and I don't know what to do with that:

```rust
struct HackRange<'a> {
    start: &'a [u8],
    end: &'a [u8],
}

impl<'a> AsRef<[u8]> for HackRange<'a> {
    fn as_ref(&self) -> &[u8] {
        // huh? this IS hit!
        todo!()
    }
}

impl<'a> RangeBounds<[u8]> for HackRange<'a> {
    fn start_bound(&self) -> std::ops::Bound<&[u8]> {
        Bound::Included(self.start)
    }

    fn end_bound(&self) -> std::ops::Bound<&[u8]> {
        Bound::Excluded(self.end)
    }
}
```

**Describe the solution you'd like**

It would be nice to have some kind of manual range function that I could use for this, as I'd like to use it for pagination in a rest API. I sort of want something between `prefix` and `range` for this.

**Describe alternatives you've considered**

I could do manual iteration over the whole prefix, which is not *great*, but maybe fast enough. It feels like I'm probably holding something wrong though!

**Additional context**

**Comment by marvin-j97** [comment, marked as answer]
`range("devices/123/logs/000000AF"..="devices/123/logs/000000BD")` should work fine.

```rs
fn main() -> fjall::Result<()> {
    let keyspace = fjall::Config::default().open()?;
    let partition = keyspace.open_partition("default", Default::default())?;

    partition.insert("devices/122/logs/000000AE", "!!!")?;
    partition.insert("devices/123/logs/000000AE", "!!!")?;
    partition.insert("devices/123/logs/000000AF", "inside")?;
    partition.insert("devices/123/logs/000000B0", "inside")?;
    partition.insert("devices/123/logs/000000BD", "inside")?;
    partition.insert("devices/123/logs/000000BE", "!!!")?;
    partition.insert("devices/124/logs/000000AE", "!!!")?;

    assert_eq!(
        3,
        partition
            .range("devices/123/logs/000000AF"..="devices/123/logs/000000BD")
            .count()
    );

    Ok(())
}
```

**Reply by jamesmunns** [reply]
Haha, just kidding, I copy and pasted another function and was still calling `.prefix` instead of `.range`.

Sorry for the noise.

**Comment by jamesmunns** [comment]
Hmm, let me try to make a better repro. My actual data here is `[u8]` (it's serialized binary data), let me see if I can can replicate.
